File: physics/auto_m.py
import math,time
import random

class Body:
    def __init__(self,mass,init_position_x,init_position_y,init_velocity_x,init_velocity_y,force_x,force_y):
        self.mass = mass
        self.position = [init_position_x,init_position_y]
        self.velocity = [init_velocity_x,init_velocity_y]
        self.force = [force_x , force_y]
        self.path_x = []
        self.path_y = []
def generate(n,mass,pos_x,pos_y,vel_x,vel_y):  
        bodies = []
        for i in range(n):
            mass =random.uniform(1,mass)
            x = random.uniform(-pos_x,pos_x)
            y = random.uniform(-pos_y,pos_y)
            vx = random.uniform(-vel_x,vel_x)
            vy = random.uniform(-vel_y,vel_y)

            bodies.append(Body(mass,x,y,vx,vy,0,0))
        return bodies           
                               

bodies = generate(5,10000,250,250,20,20)
g = 1
t = 0.01
def calc(g,bodies,t,eps):   
    for body in bodies:
        body.force[0] = 0
        body.force[1] = 0 
    for i in range(len(bodies)):
        for j in range(len(bodies)):   
            if i == j:
                continue 
            r = math.sqrt(( ( bodies[i].position[0] - bodies[j].position[0] ) ** 2 ) + ( ( bodies[i].position[1] - bodies[j].position[1] ) ** 2 ) + eps ** 2)
            d = (bodies[i].position[0]-bodies[j].position[0] , bodies[i].position[1]-bodies[j].position[1])
            F = g * (bodies[i].mass * bodies[j].mass) / r ** 3
            bodies[i].force[0] += d[0] * -F 
            bodies[i].force[1] += d[1] * -F
           

    for body in bodies:
        body.acceleration = (body.force[0]/body.mass, body.force[1]/body.mass)

        body.velocity[0] += (body.acceleration[0] * t)
        body.velocity[1] += (body.acceleration[1] * t) 

        body.position[0] += (body.velocity[0] * t)
        body.position[1] += (body.velocity[1] * t)  

        body.path_x.append(body.position[0])
        body.path_y.append(body.position[1])
def simulate(bodies,n):
    g = 1
    t = 0.001
    for _ in  range(n):
        calc(g,bodies,t,0.0001)
    print("done")

File: physics/auto-m.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <random>
using namespace std;
struct Body{
    float mass;
    float pos_x;
    float pos_y;
    float vel_x;
    float vel_y;
    float force_x;
    float force_y;
    vector <float> acceleration;
    vector <float> path_x;
    vector <float> path_y;

};
vector <Body> degenerate(float n,float mass,float pos_x,float pos_y,float vel_x,float vel_y){
    vector <Body> bodies;
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<float> massS(0,mass);
    uniform_real_distribution<float> pos_X(-pos_x,pos_x);
    uniform_real_distribution<float> pos_Y(-pos_y,pos_y);
    uniform_real_distribution<float> vx(-vel_x,vel_x);
    uniform_real_distribution<float> vy(-vel_y,vel_y);
    for (int i = 0; i < n; i++){
    float massa = massS(gen);
    float x = pos_X(gen);
    float y = pos_Y(gen);
    float vpx = vx(gen);
    float vpy = vy(gen);
    
    bodies.push_back({massa,x,y,vpx,vpy});
    }
    return bodies;
}
void save_to_file(const vector<Body>& bodies) {
    ofstream file("output.csv");

    for (int i = 0; i < bodies.size(); i++) {
        for (int j = 0; j < bodies[i].path_x.size(); j++) {
            file << j << ","      // step
                 << i << ","      // body id
                 << bodies[i].path_x[j] << ","
                 << bodies[i].path_y[j] << "\n";
        }
    }

    file.close();
}
void calc(float g, vector <Body>&arr,float t,float eps){
    for (int i = 0; i < arr.size();i++){
        arr[i].force_x = 0;
        arr[i].force_y = 0; 
    }
    for(int i = 0; i < arr.size();i++){
        for(int j = 0; j < arr.size();j++){
            if (i == j){
                continue;
            }
            float r = sqrt(pow(arr[i].pos_x - arr[j].pos_x,2) + pow(arr[i].pos_y - arr[j].pos_y,2)+ pow(eps,2));
            float d[2] = {arr[i].pos_x-arr[j].pos_x , arr[i].pos_y-arr[j].pos_y};
            float F = g * (arr[i].mass * arr[j].mass)/(pow(r,3));
            arr[i].force_x += d[0] * -F;
            arr[i].force_y += d[1] * -F;
    }
    }
    for(int i = 0; i < arr.size();i++){
        arr[i].acceleration.resize(2);
        arr[i].acceleration[0] = arr[i].force_x/arr[i].mass;
        arr[i].acceleration[1] = arr[i].force_y/arr[i].mass;
        
        arr[i].vel_x += (arr[i].acceleration[0]*t);
        arr[i].vel_y += (arr[i].acceleration[1]*t);

        arr[i].pos_x += (arr[i].vel_x * t);
        arr[i].pos_y += (arr[i].vel_y * t);

        arr[i].path_x.push_back(arr[i].pos_x);
        arr[i].path_y.push_back(arr[i].pos_y);
    }
}
int main(){

    
    vector<Body> b = degenerate(5,10000,250,250,20,20);
    
    for (int i = 0; i <= 10000 ; i++ ){
    calc(1,b,0.01,0.001);

    }
    cout<<b.size()<<endl;
    save_to_file(b);
    return 0;
}
